src/Umbraco.Core/Runtime/SqlMainDomLock.cs

```diff
@@ -0,0 +1,407 @@
+﻿using System;
+using System.Data;
+using System.Data.SqlClient;
+using System.Diagnostics;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Umbraco.Core.Logging;
+using Umbraco.Core.Persistence;
+using Umbraco.Core.Persistence.Dtos;
+using Umbraco.Core.Persistence.Mappers;
+using Umbraco.Core.Persistence.SqlSyntax;
+
+namespace Umbraco.Core.Runtime
+{
+    internal class SqlMainDomLock : IMainDomLock
+    {
+        private string _lockId;
+        private const string MainDomKey = "Umbraco.Core.Runtime.SqlMainDom";
+        private readonly ILogger _logger;
+        private IUmbracoDatabase _db;
+        private CancellationTokenSource _cancellationTokenSource = new CancellationTokenSource();
+        private SqlServerSyntaxProvider _sqlServerSyntax = new SqlServerSyntaxProvider();
```

Comment:
I think we need a bit more smooth handing if you are setting
```
<add key="Umbraco.Core.MainDom.Lock" value="SqlMainDomLock" />
```
while running SqlCE. It just explodes with `System.TimeoutException: Cannot acquire MainDom`
But the underlying issue is because SqlCE doesn't understand the Query Hints.

I discussed with @clausjensen, whether this needs to support SqlCe 
While I can't see any technical reasons it can't, we both agreed on it is okay if it just is a clear error message. You should never run `SqlMainDomLock` and `SqlCe` together.

For now, I suggest ensuring the DB is running SqlServer, otherwise throw an NotSupportedException.

Reply:
Yep, there's no reason why this should be used with SqlCe and it wasn't written with an intention to do so. Will fix up code to throw.